### PR TITLE
Refactored geoNear to use models from the offset.

### PR DESCRIPTION
**Pull request by ryanong**
Okay this was a huge refactor.
Removed to_models, to_model because they are all models to begin with now.
I was getting huge errors when using to_models and to_model with paginate.

However asc and dsc doesn't work anymore.
I kind of want to try to get this to work as a criteria but that gets kind of messy. I tend to like less magic. So maybe add another criteria options.

I have some opinions on using the

``` ruby
self.class.send :field, :distance, :type => Float
self.class.send :field, :fromHash, :type => String
self.class.send :field, :fromPoint, :type => Array
```

this means if the developer uses save on these models it saves fromPoint, fromHash and distance to the database. You might want to just make them attr_accessors I feel that this is safer over all. Distances should never really be saved to the database.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -4,14 +4,20 @@ source "http://rubygems.org"
 # gem 'ruby-debug'
 # gem 'ruby-debug19'
 
-gem 'mongoid',        '> 2'
-gem "bson",           '>= 1.3' # for non jruby apps, require bson_ext in your Gemfile to boost performance
-gem 'activesupport',  '> 3'
-gem 'hashie',         '>= 0.4.0'   # https://github.com/okiess/mongo-hashie ???
```
Comment (ryanong) [inline Gemfile:10]: No longer need hashie because we are creating models already

File: lib/mongoid/geo/geo_near.rb [hunk 2]
```diff
@@ -25,99 +24,52 @@ module Mongoid
...
+        results = (results['results'].kind_of?(Array)&& results['results'].size > 0) ? results['results'].map do |qr|
+          hash = qr['obj'].to_hash
+          res = klass.instantiate(hash) # loading up data from db with parsing
+          res.new_record = false # not new record at all.
+          ids << res.id
           res.fromPoint = qr['fromPoint']
-          # res.fromHash = qr['fromHash']          
-          res.distance = qr[distance_meth]
+          res.fromHash = qr['fromHash']
+          res.distance = (Mongoid::Geo.mongo_db_version >= 1.7) ? qr['dis'] : calc_distance(qr, center, location_attribute, options)
```
Comment (ryanong) [inline lib/mongoid/geo/geo_near.rb:65]: We should make these accessors not fields

File: lib/mongoid/geo/geo_near.rb [hunk 3]
```diff
@@ -25,99 +24,52 @@ module Mongoid
...
 
           # mongodb < 1.7 returns degrees but with earth flat. in Mongodb 1.7 you can set sphere and let mongodb calculate the distance in Miles or KM
           # for mongodb < 1.7 we need to run Haversine first before calculating degrees to Km or Miles. See below.
           near_query["distanceMultiplier"]  = distance_multiplier(options)
-          near_query["query"]               = query if query
+          near_query["query"]               = options[:query] if options[:query]
 
           # works in mongodb 1.7 but still in beta and not supported by mongodb
-          near_query["spherical"]  = true if mode == :sphere && Mongoid::Geo.mongo_db_version >= 1.7
+          near_query["spherical"]  = (options[:spherical] && Mongoid::Geo.mongo_db_version >= 1.7)
```
Comment (ryanong) [inline lib/mongoid/geo/geo_near.rb:50]: When creating API's for other API's try to keep continuity in specs if the given API is not a piece of crap.
MongoDBs api for it is fine mind as well follow it.

**Comment by kristianmandrup** [comment]
"this means if the developer uses save on these models it saves fromPoint, fromHash and distance to the database. You might want to just make them attr_accessors I feel that this is safer over all. Distances should never really be saved to the database."

Thanks. I totally agree. It was just a POC, but I never got around to make it more elegant once it "worked". I wasn't sure if Mongoid criteria etc. cold work on attr_accessors or only on fields. Will try ;) 

Thank you so much for your awesome work and contribution :) Cheers!

File: lib/mongoid/geo/geo_near.rb [hunk 4]
```diff
@@ -25,99 +24,52 @@ module Mongoid
...
-      def to_criteria
-        ids = map(&:_id)  
-        Mongoid::Criteria.new(first.klass).where(:_id.in => ids).desc(:distance)
-      end      
-    end
-
     module Near
       def geoNear(center, location_attribute, options = {})
-        center = center.respond_to?(:collection) ? eval("center.#{location_attribute}") : center
+        center = (center.respond_to?(:collection) && center.respond_to?(location_attribute)) ? center.public_send(location_attribute) : center
```
Comment (ryanong) [inline lib/mongoid/geo/geo_near.rb:29]: Never use eval if you can avoid it.
Maybe change this to

```
if center.respond_to?(:collection)
  if center.respond_to?(location_attribute))
    center = center.public_send(location_attribute)
  elsif center[location_attribute]
    center = center[location_attribute]
  end
end
```

File: lib/mongoid/geo/geo_near.rb [hunk 5]
```diff
@@ -25,99 +24,52 @@ module Mongoid
...
           res.fromPoint = qr['fromPoint']
-          # res.fromHash = qr['fromHash']          
-          res.distance = qr[distance_meth]
+          res.fromHash = qr['fromHash']
+          res.distance = (Mongoid::Geo.mongo_db_version >= 1.7) ? qr['dis'] : calc_distance(qr, center, location_attribute, options)
           res._id = qr['obj']['_id'].to_s
           res
-        end          
+        end : []
+        Mongoid::Criteria.new(klass).where(:_id.in => ids)
```
Comment (ryanong) [inline lib/mongoid/geo/geo_near.rb:69]: delete this line woops

File: Rakefile [hunk 6]
```diff
@@ -21,13 +21,6 @@ Jeweler::Tasks.new do |gem|
   gem.description = %Q{Makes it easy to use geo calculations with Mongoid}
   gem.email = "kmandrup@gmail.com"
   gem.authors = ["Kristian Mandrup"]
-  # Include your dependencies below. Runtime dependencies are required when using your gem,
-  # and development dependencies are only needed for development (ie running rake tasks, tests, etc)
-  gem.add_runtime_dependency      'mongoid',        '> 2'
-  gem.add_runtime_dependency      'bson',           '>= 1.3' 
-  gem.add_runtime_dependency      'activesupport',  '> 3'
-  gem.add_runtime_dependency      'hashie',         '>= 0.4.0'   # https://github.com/okiess/mongo-hashie ???  
-  gem.add_development_dependency  'rspec',          '> 2.4'
```
Comment (ryanong) [inline Rakefile:30]: This info is automatically pulled from the Gemfile by jeweler

File: Gemfile [hunk 7]
```diff
@@ -4,14 +4,20 @@ source "http://rubygems.org"
...
+  gem 'rspec',    '>=2.4'
+  gem 'bundler',  '>=1'
+  gem 'jeweler',  '>=1.5'
+  gem 'rdoc',     '>=3.6'
+end
+
+group :test do
+  gem 'rails',        '>=3.0'
+  gem 'rspec-rails',  '>=2.6'
+  gem 'capybara',     '>=0.4'
```
Comment (ryanong) [inline Gemfile:22]: Tests don't work if this is not in gem file

**Pull request merged**